[jit] make RRef type annotation available in Python

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33263 [jit] allow RRef local creation with IValue objects
* #33582 [rpc] special case tensor type check when getting RRef
* **#33526 [jit] make RRef type annotation available in Python**
* #33369 [jit] infer RRef type as container type

This PR make RRef type annotation available in Python by introducing a
fake generic type which holds the element type, it allows user use rref
as a type annotation in python functions that jit calls into (i.e.
jit.ignore)

Differential Revision: [D19988848](https://our.internmc.facebook.com/intern/diff/D19988848)

wanchaol committed:
[jit] make RRef type annotation available in Python

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] make RRef type annotation available in Python"




[ghstack-poisoned]

wanchaol committed:
Update on "[jit] make RRef type annotation available in Python"


This PR make RRef type annotation available in Python by introducing a
fake generic type which holds the element type, it allows user use rref
as a type annotation in python functions that jit calls into (i.e.
jit.ignore)

Differential Revision: [D19988848](https://our.internmc.facebook.com/intern/diff/D19988848)

[ghstack-poisoned]

driazati commented on torch/_jit_internal.py:
```diff
@@ -663,6 +663,37 @@ def is_final(ann):
+        return getattr(ann, "__origin__", None) is RRef
+
+except ImportError:
+    class RRefInstance(object):
```
`typing` should always be installed alongside PyTorch so I think you can get rid of the whole try catch and `RRefInstance`

wanchaol replied:
I think it's still not installed in some CI, which is blocked me landing https://github.com/pytorch/pytorch/pull/27637, I will figure out how to fix that CI and submit a separate PR to remove the whole try catch along with other cases.

driazati reviewed (approved):
lgtm, can you add a test that code that uses this type at least compiles?

wanchaol committed:
Update on "[jit] make RRef type annotation available in Python"


This PR make RRef type annotation available in Python by introducing a
fake generic type which holds the element type, it allows user use rref
as a type annotation in python functions that jit calls into (i.e.
jit.ignore)

Differential Revision: [D19988848](https://our.internmc.facebook.com/intern/diff/D19988848)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] make RRef type annotation available in Python"


This PR make RRef type annotation available in Python by introducing a
fake generic type which holds the element type, it allows user use rref
as a type annotation in python functions that jit calls into (i.e.
jit.ignore)

Differential Revision: [D19988848](https://our.internmc.facebook.com/intern/diff/D19988848)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] make RRef type annotation available in Python"


This PR make RRef type annotation available in Python by introducing a
fake generic type which holds the element type, it allows user use rref
as a type annotation in python functions that jit calls into (i.e.
jit.ignore)

Differential Revision: [D19988848](https://our.internmc.facebook.com/intern/diff/D19988848)

[ghstack-poisoned]
